ext/crates/fp/src/field/smallfq.rs

```diff
@@ -0,0 +1,259 @@
+use std::sync::{Arc, LazyLock};
+
+use dashmap::DashMap as HashMap;
+
+use super::{
+    element::{FieldElement, FieldElementContainer},
+    field_internal::FieldInternal,
+    Field, Fp,
+};
+use crate::{
+    constants::BITS_PER_LIMB,
+    limb::Limb,
+    prime::{log2, Prime, ValidPrime},
+    vector::inner::FqVector,
+    PRIME_TO_INDEX_MAP,
+};
+
+static SMALL_CONWAY_POLYS: [[[u32; 17]; 15]; 54] = include!("small_conway_polys.txt");
+
+type ZechTable = HashMap<SmallFqElement, SmallFqElement>;
+type Polynomial<P> = FqVector<Fp<P>>;
+
+/// A table of lazily initialized [Zech logarithms][zech_logs].
+///
+/// Key is the field, value is a fully initialized table of Zech logarithms.
+///
+/// [zech_logs]: https://en.wikipedia.org/wiki/Zech%27s_logarithm
+static ZECH_LOGS: LazyLock<HashMap<(ValidPrime, u32), Arc<ZechTable>>> =
+    LazyLock::new(HashMap::new);
```

Comment:
I like the idea of adding a `&'static ZechTable` to the `Field`. Unfortunately the static has to be wrapped in `LazyLock` because `HashMap::new` isn't const, but `LazyLock<HashMap<(ValidPrime, u32), LazyLock<ZechTable>>>` would work. I'll experiment with that

Reply:
What about a `LazyLock<Vec<LazyLock<ZechTable>>>` and a `LazyLock<Vec<usize>>`? Can you get an `&'static ZechTable` from that or does it have the same problem?

Reply:
I think the best option is to just construct the table and then `Box::leak` it. I gave it a go in [b8d58fe](https://github.com/SpectralSequences/sseq/pull/159/commits/b8d58fe0c2a515c9062b665324cbcdd9beb81f88)

Reply:
Here's a flamegraph of adding different numbers in a loop:
```rs
#[bench]
fn bench_f16_add(b: &mut Bencher) {
    let f16 = SmallFq::new(P2, 4);
    let zero = f16.zero();
    let one = f16.one();
    let a = f16.a();

    let mut elements = vec![zero, one];
    for _ in 1..f16.q() {
        let prev = elements.last().unwrap();
        elements.push(*prev * a);
    }
    b.iter(|| {
        for i in 0..15 {
            for j in 0..15 {
                test::black_box(elements[i] + elements[j]);
            }
        }
    });
}
```

I think 43% of the execution time is spent in the outer hash lookup of the ZechTable and 57% of the time is spent in the inner hash lookup. Essentially no time is spent doing the actual work of the addition. 

![flame](https://github.com/user-attachments/assets/01743f95-14be-49d5-9673-254d1513e7ad?x=20482471&y=405)

Reply:
If I get rid of the outer hash lookup execution time goes down from 15,760.17 ns/iter to 8,895.13 ns/iter, a 43.5% reduction
```patch
--- a/ext/crates/fp/src/field/smallfq.rs
+++ b/ext/crates/fp/src/field/smallfq.rs
@@ -1,4 +1,4 @@
-use std::sync::{Arc, LazyLock};
+use std::sync::{Arc, LazyLock, OnceLock};
 
 use dashmap::DashMap as HashMap;
 
@@ -25,13 +25,12 @@ type Polynomial<P> = FqVector<Fp<P>>;
 /// Key is the field, value is a fully initialized table of Zech logarithms.
 ///
 /// [zech_logs]: https://en.wikipedia.org/wiki/Zech%27s_logarithm
-static ZECH_LOGS: LazyLock<HashMap<(ValidPrime, u32), Arc<ZechTable>>> =
-    LazyLock::new(HashMap::new);
+static ZECH_LOGS: OnceLock<ZechTable> = OnceLock::new();
 
 /// Return the Zech logarithm table for the given field. If it does not exist yet, initialize it.
 /// The initialization might be fairly expensive (several ms).
-fn zech_logs<P: Prime>(fq: SmallFq<P>) -> Arc<ZechTable> {
-    let table = ZECH_LOGS.entry((fq.p.to_dyn(), fq.d)).or_insert_with(|| {
+fn zech_logs<P: Prime>(fq: SmallFq<P>) -> &'static ZechTable {
+    ZECH_LOGS.get_or_init(|| {
         let prime_field = Fp::new(fq.characteristic());
         let conway_poly = {
             let v = SMALL_CONWAY_POLYS[PRIME_TO_INDEX_MAP[fq.p.as_usize()]][fq.d as usize - 2]
@@ -85,9 +84,8 @@ fn zech_logs<P: Prime>(fq: SmallFq<P>) -> Arc<ZechTable> {
 
             current = mul_by_a(current);
         }
-        Arc::new(table)
-    });
-    Arc::clone(&table)
+        table
+    })
 }
 
 /// A field of order `q = p^d`, where `q < 2^16` and `d > 1`. Fields of that size are small enough
```

Reply:
I see what you mean. Storing a &'static to a leaked table should clear the first hash completely. The second hash is unavoidable though, and doing addition really *is* that lookup, arguably. I'll see if I can add a bench for `FqVector`s over those fields

Reply:
If I get rid of the other hash map too:
```patch
diff --git a/ext/crates/fp/src/field/smallfq.rs b/ext/crates/fp/src/field/smallfq.rs
index 78cbb5bd..c90430c6 100644
--- a/ext/crates/fp/src/field/smallfq.rs
+++ b/ext/crates/fp/src/field/smallfq.rs
@@ -17,7 +17,7 @@ use crate::{
 
 static SMALL_CONWAY_POLYS: [[[u32; 17]; 15]; 54] = include!("small_conway_polys.txt");
 
-type ZechTable = HashMap<SmallFqElement, SmallFqElement>;
+type ZechTable = Vec<SmallFqElement>;
 type Polynomial<P> = FqVector<Fp<P>>;
 
 /// A table of lazily initialized [Zech logarithms][zech_logs].
@@ -70,24 +69,17 @@ fn zech_logs<P: Prime>(fq: SmallFq<P>) -> Arc<ZechTable> {
         }
 
         // Loop over all elements again, but now recording logarithms.
-        let table = ZechTable::new();
-        table.insert(fq.zero().val(), fq.one().val());
-
+        let mut table = vec![SmallFqElement(None); (fq.q() - 1) as usize];
         let mut current = Polynomial::new(prime_field, conway_poly.len());
         current.set_entry(0, prime_field.one());
         for i in 0..fq.q() - 1 {
             let mut current_plus_1 = current.clone();
             current_plus_1.add_basis_element(0, prime_field.one());
-            table.insert(
-                SmallFqElement(Some(i)),
-                SmallFqElement(poly_to_power.get(&current_plus_1).as_deref().copied()),
-            );
-
+            table[i as usize] = SmallFqElement(poly_to_power.get(&current_plus_1).as_deref().copied());
             current = mul_by_a(current);
         }
-        Arc::new(table)
-    });
-    Arc::clone(&table)
+        table
+    })
 }
 
 /// A field of order `q = p^d`, where `q < 2^16` and `d > 1`. Fields of that size are small enough
@@ -166,7 +158,7 @@ impl<P: Prime> FieldInternal for SmallFq<P> {
                 // a^m + a^n = a^m (1 + a^(n - m)) = a^(m + Zech(n - m))
                 let table = zech_logs(self);
                 let (a, b) = if a >= b { (a, b) } else { (b, a) };
-                let zech = table.get(&SmallFqElement(Some(a - b))).unwrap();
+                let zech = table[(a - b) as usize];
                 if let Some(zech) = zech.0 {
                     SmallFqElement(Some(b + zech))
                 } else {
```
Execution time goes down to 745.14 ns/iter. So indeed, 95% of the original execution time is spent in hash map lookups.

Reply:
You're absolutely right. I changed the code so that fields store 'static references and tables are stored as `Vec`s and I got a similar value of 625 ns for that double loop.

In fact, writing a Criterion + pprof benchmark I also saw that computing q all the time was a major bottleneck. Caching it makes this bench go down to 210 ns, and a similar bench that tests multiplication goes from 436 ns to just 5.3 ns!

Thanks for insisting. The code looks better too!